feat(compiler-cli): no longer re-export external symbols by default

devversion opened the pull request:
With #28594 we refactored the `@angular/compiler` slightly to
allow opting out from external symbol re-exports which are
enabled by default.

Since symbol re-exports only benefit projects which have a
very strict dependency enforcement, external symbols should
not be re-exported by default as this could grow the size of
factory files and cause unexpected behavior with Angular's
AOT symbol resolving (e.g. see: #25644).

Additionally "ngtsc" also does not provide any way of using
external symbol re-exports, so this means that NGC partially
matches the behavior of "ngtsc" then (unless explicitly opted-out).

Internally at Google, symbol re-exports need to be still enabled
since Google has a very strict dependency enforcement (even of
produced JavaScript output), so the `ng_module` Bazel rule will
enable the symbol re-exports by default when running within Blaze.

Fixes #25644.

devversion commented:
@IgorMinar @alxhub @alexeagle This is just the change from the other PR but we split the changes into two parts (see [here](https://github.com/angular/angular/pull/28594#pullrequestreview-201816956)). Can you review/re-approve this PR. It will be still blocked until #28594 is merged.

**Note** that this change needs some test updates because the symbol re-exports will be _disabled_ by default. All of these test changes seem reasonable and I've also added more test cases with the new expected behavior.

The diff just looks huge, but if you disable whitespace, it should be very reduced and clear.

IgorMinar reviewed (commented):
Can you please document why the strictness guarantees that blaze enforces doesn't apply to bazel? (I actually don't know why and it seems a bit suspicious - @alexeagle is this right?)

alexeagle commented:
I don't know if it's right, it surprised me too, I think Alex suggested it. If anything it might be a diff between google3 and other repos rather than between bazel and blaze

devversion commented:
@IgorMinar I've documented this better in the commit description and in the comments. This is based on what @alxhub wrote [here](https://github.com/angular/angular/pull/28594#discussion_r255259724). Please have another look.

alxhub commented:
@IgorMinar @alexeagle correct, it's because in g3 we have AJD which checks the generated code in certain circumstances, whereas in most repos we only check strictdeps at the TS level, before emit.

alxhub commented on packages/bazel/src/ng_module.bzl:
```diff
@@ -245,6 +245,10 @@ def _ngc_tsconfig(ctx, files, srcs, **kwargs):
         "enableIvy": _enable_ivy_value(ctx),
         "fullTemplateTypeCheck": ctx.attr.type_check,
+        # In Blaze we still want to use the symbol factory re-exports in order to
+        # not break existing apps inside google3. Enabling this in google3 would cause
```
```suggestion
        # not break existing apps inside google3. Disabling this in google3 would cause
```

devversion replied:
Updated. thanks!

alexeagle commented:
[presubmit](http://test/OCL:233647421:BASE:233647763:1550000426041:7ab2488e)

devversion committed:
feat(compiler-cli): no longer re-export external symbols by default

With #28594 we refactored the `@angular/compiler` slightly to
allow opting out from external symbol re-exports which are
enabled by default.

Since symbol re-exports only benefit projects which have a
very strict dependency enforcement, external symbols should
not be re-exported by default as this could grow the size of
factory files and cause unexpected behavior with Angular's
AOT symbol resolving (e.g. see: #25644).

Note that the common strict dependency enforcement for source
files does still work with external symbol re-exports disabled,
but there are also strict dependency checks that enforce strict
module dependencies also for _generated files_ (such as the
ngfactory files). This is how Google3 manages it's dependencies
and therefore external symbol re-exports need to be enabled within
Google3.

Also "ngtsc" also does not provide any way of using external symbol
re-exports, so this means that with this change, NGC can partially
match the behavior of "ngtsc" then (unless explicitly opted-out).

As mentioned before, internally at Google symbol re-exports need to
be still enabled, so the `ng_module` Bazel rule will enable the symbol
re-exports by default when running within Blaze.

Fixes #25644.

IgorMinar commented:
http://test/OCL:233688736:BASE:233688759:1550037327974:d00a37fc

IgorMinar commented:
merge-assistance: please check on the presubmit result once it's ready. this PR looks good to me.
